app/controllers/single_pages_controller.rb

```diff
@@ -67,11 +67,11 @@ def download_samples_excel
 
     @study = Study.find(study_id)
     @project = @study.projects.first
-    if sample_ids.nil? || sample_ids == [] || sample_type_id.nil?
-      raise "Nothing to export to Excel."
-    end
-    @samples = Sample.where(id: sample_ids)
+    raise "Nothing to export to Excel." if sample_ids.nil? || sample_ids == [] || sample_type_id.nil?
+
+    @samples = sample_ids.map { |sample_id| Sample.find(sample_id) if Sample.find(sample_id).can_view? }
```

Comment:
> bear in mind Sample.where(id: sample_ids) may not preserve the order, so you might want to reorder the samples again afterwards, but with a large number of samples it will still be much quicker

I didn't know that. I changed it to:

```ruby
@samples = Sample.where(id: sample_ids)&.authorized_for(:view).sort_by(&:id)
```

It looks a lot cleaner now too.